applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -919,6 +937,109 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::ComputePotentia
     }
 }
 
+template <int TDim, int TNumNodes>
+void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::FindUpwindElement(const ProcessInfo& rCurrentProcessInfo)
+{
+    GeometryType upwind_element_boundary;
+    FindUpwindEdge(upwind_element_boundary, rCurrentProcessInfo);
+    std::vector<size_t> upwind_element_nodes;
+    PotentialFlowUtilities::GetSortedIds<TDim, TNumNodes>(upwind_element_nodes, upwind_element_boundary);
+
+    GlobalPointersVector<Element> upwind_element_candidates;
+    PotentialFlowUtilities::GetNodeNeighborElementCandidates<TDim, TNumNodes>(upwind_element_candidates, upwind_element_boundary);
+    SelectUpwindElement(upwind_element_nodes, upwind_element_candidates);
+}
+
+template <int TDim, int TNumNodes>
+void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::FindUpwindEdge(GeometryType& rUpwindEdge,
+    const ProcessInfo& rCurrentProcessInfo)
+{   
+
+    GeometriesArrayType element_boundary_geometry;
+    GetElementGeometryBoundary(element_boundary_geometry);
+
+    // free stream values
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+
+    double minimum_free_stream_projection = 0.0;
+    for (SizeType i = 0; i < element_boundary_geometry.size(); i++)
+    {
+        const auto unit_normal = GetEdgeUnitNormal(element_boundary_geometry[i]);
+
+        const double free_stream_projection = inner_prod(unit_normal, free_stream_velocity);
+
+        if(free_stream_projection < minimum_free_stream_projection)
+        {
+            minimum_free_stream_projection = free_stream_projection;
+            rUpwindEdge = element_boundary_geometry[i];
+        }
+    }
+}
+
+template<int TDim, int TNumNodes>
+void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::GetElementGeometryBoundary(GeometriesArrayType& rElementGeometryBoundary)
+{
+    const TransonicPerturbationPotentialFlowElement& r_this = *this;
+    
+    // current element geometry
+    const GeometryType& r_geom = r_this.GetGeometry();
+
+    // get element edges or faces depending on dimension of the problem
+    if(TDim == 2)
+    {
+        // current element edges
+        rElementGeometryBoundary = r_geom.GenerateEdges();
+    }
+    else if(TDim == 3)
+    {
+        // current element faces
+        rElementGeometryBoundary = r_geom.GenerateFaces();
+    }
+}
+
+template <int TDim, int TNumNodes>
+array_1d<double, 3> TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::GetEdgeUnitNormal(const GeometryType& rEdge)
+{
+    // get local coordinates of edge center
+    array_1d<double, 3> edge_center_coordinates;
+    rEdge.PointLocalCoordinates(edge_center_coordinates, rEdge.Center());
+
+    // outward pointing normals of each edge
+    const auto normal = rEdge.Normal(edge_center_coordinates);
+    return normal / std::sqrt(inner_prod(normal, normal));
```

Comment:
removing normalization

```suggestion
    return rEdge.Normal(edge_center_coordinates);
```